keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,537 @@
+  // +optional
+  MinDomains *int32
+}
+```
+
+### Implementation details
+
+**DoNotSchedule implementation**
+
+In Filter of Pod Topology Spread, current filtering criteria is
+
+```
+('existing matching num' + 'if self-match (1 or 0)' - 'global min matching num') <= 'maxSkew'
+```
+
+- `existing matching num` denotes the number of current existing matching Pods on the domain.
+- `if self-match` denotes if the labels of Pod is match with selector of the constraint.
+- `global min matching num` denotes the minumun number of matching Pods.
+
+For `whenUnsatisfiable: DoNotSchedule`, Pod Topology Spread will treat `global min matching num` as 0
+when the number of domains with matching topology keys is less than `minDomains`.
+
+We can calculate the number of domains with matching topology keys in PreFilter, along with the calculation of [`TpPairToMatchNum`](https://github.com/kubernetes/kubernetes/blob/0153febd9f0098d4b8d0d484927710eaf899ef40/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go#L49).
+This extra calculation doesn't increase the complexity of the preFilter logic.
+Pod Topology Spread will be able to use the number of domains to determine the value of `global min matching num` when we calculate filtering criteria.
+
+**ScheduleAnyway implementation**
+
+When the number of domains with matching topology keys is less than `minDomains` and `whenUnsatisfiable` equals to `ScheduleAnyway`,
+Pod Topology Spread will give low scores to Nodes on the domains which have the same or more number of matching Pods as `maxSkew`.
+(It is a low score, not always zero.)
+
+In Pod Topology Spread, the higher the score from Score, the lower will be the final score calculated by Normalized Score.
+So, Pod Topology Spread should give high scores to non-preferred Nodes in Score.
+
+In Score, the score for each constraint is evaluated, and the sum of those scores will be a Node score.
+
+Basically, the score for each constraint is evaluated this way:
+
+```
+'existing matching num' * 'topology weight' + 'maxSkew' - 1
+```
+
+When the number of domains with matching topology keys is less than `minDomains`,
+Pod Topology Spread doubles that score (so that final score will be a low score) if this criteria is met:
+
+```
+('existing matching num' + 'if self-match (1 or 0)' - 'global min matching num') > 'maxSkew'
+```
+
+- `existing matching num` denotes the number of current existing matching Pods on the domain.
+- `if self-match` denotes if the labels of Pod is match with selector of the constraint.
+- `global min matching num` denotes the minumun number of matching Pods.
+
+Like preFilter, we need to calculate the number of domains with matching topology keys and the minimum number of matching Pods in preScore,
+so that Pod Topology Spread can determine the evaluation way with them.
+
+This extra calculation may affect the performance of the preScore, because the current preScore only see Nodes which have passed the Filter.
```

Comment:
Uhm.... we need to make sure to do it only if a constraint has minDomains set.

Is it really worth though? Since the original formula already gives a similar semantic, and because scoring is ignored by cluster-autoscaler, I now think we should leave scores unmodified.

@Huang-Wei, thoughts?

Reply:
Well.. I totally agree with "the original formula already gives a similar semantic". But, isn't it strange for users that there is no score change even after setting minDomains? 

> we need to make sure to do it only if a constraint has minDomains set.

Agree, we should do that if we go the way of changing scores.

Reply:
Yes, we can specify in the documentation that it only applies to `DoNotSchedule`.

> But, isn't it strange for users that there is no score change even after setting minDomains?

The scores are not user visible regardless. The difference is so subtle that I don't think they would notice a difference in most scenarios even if there is a change.

WDYT @Huang-Wei ?

Reply:
Yes, restrict it to `DoNotSchedule` adheres to our initial motivation; while investing efforts on tuning subtle scroring difference doesn't make a lot of sense.

Reply:
Okay. Understood.
I changed the doc to set the goal to DoNotSchedule only.

Reply:
If it's not too much effort, add a summary of the discussion about DoNotSchedule in the "Alternatives" section. Explain what the proposal was, and why ultimately we decided it was not worth pursuing.

Reply:
BTW, while it might feel disappointing that we are discarding the change, it was absolutely crucial that we explored the options. So thank you a lot for your effort.

Reply:
> add a summary of the discussion about DoNotSchedule in the "Alternatives" section. Explain what the proposal was, and why ultimately we decided it was not worth pursuing.

Sure. I added it to `Alternatives`. Could you please take a look 🙏 

> BTW..

happy to contribute to our best choice :)